certbot/plugins/dns_cloudflare.py

```diff
@@ -0,0 +1,323 @@
+
+        if self.email:
+            setattr(self.config, self.dest('email'), self.email)
+        else:
+            raise errors.PluginError('Cloudflare account email address required to proceed.')
+
+        configured_api_key = self.conf('api-key')
+        if configured_api_key:
+            self.api_key = configured_api_key
+        else:
+            self.api_key = self._prompt_for_data('API key')
+
+        if self.api_key:
+            setattr(self.config, self.dest('api-key'), self.api_key)
+        else:
+            raise errors.PluginError('Cloudflare account API key required to proceed.')
+
+    @staticmethod
+    def _prompt_for_data(label):
+        """
+        Prompt the user for a piece of information.
+
+        :param string label: The user-friendly label for this piece of information.
+        :returns: The user's response (guaranteed non-empty).
+        :rtype: string
+        """
+
+        display = zope.component.getUtility(interfaces.IDisplay)
+
+        while True:
+            code, response = display.input(
+                'Input Cloudflare account {0}'.format(label),
+                force_interactive=True)
+            if code == display_util.HELP:
+                # Displaying help is not currently implemented
+                return None
+            elif code == display_util.CANCEL:
+                return None
+            else:  # code == display_util.OK
+                if not response:
+                    display.notification('Please enter an {0}.'.format(label), pause=False)
+                else:
+                    return response
+
+    def _get_cloudflare_client(self):
+        """
+        Helper method to construct a `_CloudflareClient`.
+
+        Uses configured credentials.
+
+        :return: a new
+        :rtype: `_CloudflareClient`
+        """
+
+        return _CloudflareClient(self.email, self.api_key)
+
+
+class _CloudflareClient(object):
+    """
+    Encapsulates all communication with the Cloudflare API.
+    """
+
+    def __init__(self, email, api_key):
+        self.cf = CloudFlare.CloudFlare(email, api_key)
+
+    def add_txt_record(self, domain, record_name, record_content, record_ttl):
+        """
+        Add a TXT record using the supplied information.
+
+        :param string domain: The domain to use to look up the Cloudflare zone.
+        :param string record_name: The record name (typically beginning with '_acme-challenge.').
+        :param string record_content: The record content (typically the challenge validation).
+        :param int record_ttl: The record TTL (number of seconds that the record may be cached).
+        :raises: errors.PluginError if an error occurs communicating with the Cloudflare API
+        """
+
+        zone_id = self._find_zone_id(domain)
+
+        data = {'type': 'TXT',
+                'name': record_name,
+                'content': record_content,
+                'ttl': record_ttl}
```

Comment:
I wanted to provide a little background here about dependencies in Certbot. First, I'll say that too many dependencies is one of the most common complaints we have. Some people have small lightweight servers and hate having to install 35+ packages just to get a cert. This can be mitigated by making Certbot more modular but this increases the complexity for people getting the pieces of Certbot they need for their use case. It's a trade off and we should try to strike a reasonable balance here.

Additionally, our current approach to keeping Certbot reasonably up to date as we and the ACME protocol continue to change rapidly is to try and convince Linux distros to update us more quickly than they usually do. This usually involves a lot of compromise and one instance of this is we told Debian we wouldn't add any non-optional dependencies to Certbot. None of these things are total blockers to us adding dependencies and there are definite benefits to doing so but I wanted to bring them up so we can make the right design decision here.

One question I had is about the viability of using the [lexicon](https://github.com/AnalogJ/lexicon) project you found for many of our DNS plugins including this one. All of the DNS plugins that have been implemented so far are supposedly supported by lexicon with the exception of Google Cloud DNS which they claim to be open to accepting a PR for. If we go this route, I'd also be curious if restructuring our DNS plugins makes sense if there's a common base and level of abstraction from lexicon we can use to easily support all/many of these DNS providers.

If for some reason this doesn't work, I think we have two options for this plugin. We could do both of these options but I think at least one is necessary.

1. We split it out into its own package.
2. We vendorize the dependencies that aren't widely packaged by OSes such as `cloudflare` and `logger`. I haven't looked into exactly what we'd need to vendorize here but I'm happy to do so if we go this route.

Not doing one of these two things means not only did we go back on our promise to Debian, but we now depend on libraries that aren't packaged. We can talk to our package maintainer if we really want to go this route, but I think we'll have a bad time.